File: pascal/pascal/lexer.go
package pascal

import (
	"fmt"
	"strings"
	"unicode"
)

type Lexer struct {
	text        string
	pos         int
	currentChar rune
}

func NewLexer(text string) *Lexer {
	l := &Lexer{
		text: text,
		pos:  0,
	}
	if len(text) > 0 {
		l.currentChar = rune(text[0])
	} else {
		l.currentChar = 0
	}
	return l
}

func (l *Lexer) advance() {
	l.pos++
	if l.pos >= len(l.text) {
		l.currentChar = 0
	} else {
		l.currentChar = rune(l.text[l.pos])
	}
}

func (l *Lexer) skipWhitespace() {
	for l.currentChar != 0 && unicode.IsSpace(l.currentChar) {
		l.advance()
	}
}

func (l *Lexer) integer() string {
	var result strings.Builder
	for l.currentChar != 0 && unicode.IsDigit(l.currentChar) {
		result.WriteRune(l.currentChar)
		l.advance()
	}
	return result.String()
}

func (l *Lexer) id() string {
	var result strings.Builder
	for l.currentChar != 0 && (unicode.IsLetter(l.currentChar) || unicode.IsDigit(l.currentChar) || l.currentChar == '_') {
		result.WriteRune(l.currentChar)
		l.advance()
	}
	return result.String()
}

func (l *Lexer) peek() rune {
	peekPos := l.pos + 1
	if peekPos >= len(l.text) {
		return 0
	}
	return rune(l.text[peekPos])
}

func (l *Lexer) NextToken() (Token, error) {
	for l.currentChar != 0 {
		if unicode.IsSpace(l.currentChar) {
			l.skipWhitespace()
			continue
		}

		if unicode.IsDigit(l.currentChar) {
			return NewToken(INTEGER, l.integer()), nil
		}

		if unicode.IsLetter(l.currentChar) || l.currentChar == '_' {
			idStr := l.id()
			idUpper := strings.ToUpper(idStr)
			if idUpper == "BEGIN" {
				return NewToken(BEGIN, idStr), nil
			}
			if idUpper == "END" {
				return NewToken(END, idStr), nil
			}
			return NewToken(ID, idStr), nil
		}

		if l.currentChar == ':' && l.peek() == '=' {
			l.advance()
			l.advance()
			return NewToken(ASSIGN, ":="), nil
		}

		switch l.currentChar {
		case '+':
			l.advance()
			return NewToken(PLUS, "+"), nil
		case '-':
			l.advance()
			return NewToken(MINUS, "-"), nil
		case '*':
			l.advance()
			return NewToken(MUL, "*"), nil
		case '/':
			l.advance()
			return NewToken(DIV, "/"), nil
		case '(':
			l.advance()
			return NewToken(LPAREN, "("), nil
		case ')':
			l.advance()
			return NewToken(RPAREN, ")"), nil
		case ';':
			l.advance()
			return NewToken(SEMI, ";"), nil
		case '.':
			l.advance()
			return NewToken(DOT, "."), nil
		default:
			return Token{}, fmt.Errorf("invalid character: %c", l.currentChar)
		}
	}

	return NewToken(EOF, ""), nil
}

File: interpreter_full/lexer.py
from .token import TokenType, Token

class Lexer():

    def __init__(self):
        self._text = ''
        self._pos = 0
        self._current_char = None

    def __number(self) -> str:
        result = []
        while (self._current_char is not None 
               and self._current_char.isdigit()):
            result.append(self._current_char)
            self.__forward()
        return ''.join(result)

    def next_token(self) -> Token:
        while self._current_char:
            if self._current_char.isspace(): 
                self.__skip()
                continue

            current_char = self._current_char
            if self._current_char.isdigit():
                return Token(TokenType.NUMBER, self.__number())
            if self._current_char in ['+', '-', '*', '/']:
                self.__forward()
                return Token(TokenType.OPERTATOR, current_char)
            if self._current_char == '(':
                self.__forward()
                return Token(TokenType.LPAREN, current_char)
            if self._current_char == ')':
                self.__forward()
                return Token(TokenType.RPAREN, current_char)
            raise SyntaxError()
        return Token(TokenType.EOL, "")
        
    def __skip(self):
        while (self._current_char is not None
               and self._current_char.isspace()):
            self.__forward()
    
    def __forward(self):
        self._pos += 1
        if self._pos >= len(self._text):
            self._current_char = None
        else:
            self._current_char = self._text[self._pos]
    
    def set_text(self, expr: str):
        self._text = expr
        self._pos = 0
        self._current_char = self._text[self._pos]